packages/hdwallet-provider/test/provider.test.ts

```diff
@@ -25,84 +25,204 @@ describe("HD Wallet Provider", function () {
+      }
+    });
+
+    it("provides for an array of private keys", async () => {
+      const privateKeys = [
         "3f841bf589fdf83a521e55d51afddc34fa65351161eead24f064855fc29c9580",
-      "0xbd3366a0e5d2fb52691e3e08fabe136b0d4e5929":
         "9549f39decea7b7504e15572b2c6a72766df0281cea22bd1a3bc87166b1ca290"
-    };
-
-    provider = new WalletProvider(privateKeys, `http://localhost:${port}`);
-    web3.setProvider(provider);
-
-    const addresses = provider.getAddresses();
-    assert.equal(
-      addresses.length,
-      privateKeys.length,
-      "incorrect number of wallets derived"
-    );
-    addresses.forEach(address => {
-      assert(EthUtil.isValidAddress(address), "invalid address");
-      const privateKey = Buffer.from(privateKeysByAddress[address], "hex");
-      const expectedAddress = `0x${EthUtil.privateToAddress(
-        privateKey
-      ).toString("hex")}`;
+      ];
+
+      const privateKeysByAddress: { [address: string]: string } = {
+        "0xc515db5834d8f110eee96c3036854dbf1d87de2b":
+          "3f841bf589fdf83a521e55d51afddc34fa65351161eead24f064855fc29c9580",
+        "0xbd3366a0e5d2fb52691e3e08fabe136b0d4e5929":
+          "9549f39decea7b7504e15572b2c6a72766df0281cea22bd1a3bc87166b1ca290"
+      };
+
+      provider = new HDWalletProvider({
+        privateKeys,
+        providerOrUrl: `http://localhost:${port}`
+      });
+      web3.setProvider(provider);
+
+      const addresses = provider.getAddresses();
       assert.equal(
-        address,
-        expectedAddress,
-        "incorrect address for private key"
+        addresses.length,
+        privateKeys.length,
+        "incorrect number of wallets derived"
       );
+      addresses.forEach(address => {
+        assert(EthUtil.isValidAddress(address), "invalid address");
+        const privateKey = Buffer.from(privateKeysByAddress[address], "hex");
+        const expectedAddress = `0x${EthUtil.privateToAddress(
+          privateKey
+        ).toString("hex")}`;
+        assert.equal(
+          address,
+          expectedAddress,
+          "incorrect address for private key"
+        );
+      });
+
+      const number = await web3.eth.getBlockNumber();
+      assert(number === 0);
     });
 
-    web3.eth.getBlockNumber((err, number) => {
+    it("provides for a private key", async () => {
+      const privateKey =
+        "3f841bf589fdf83a521e55d51afddc34fa65351161eead24f064855fc29c9580"; //random valid private key generated with ethkey
+      provider = new HDWalletProvider({
+        privateKey,
+        providerOrUrl: `http://localhost:${port}`
+      });
+      web3.setProvider(provider);
+
+      const addresses = provider.getAddresses();
+      assert.equal(addresses[0], "0xc515db5834d8f110eee96c3036854dbf1d87de2b");
+      addresses.forEach(address => {
+        assert(EthUtil.isValidAddress(address), "invalid address");
+      });
+
+      const number = await web3.eth.getBlockNumber();
```

Comment:
I think we should disallow the single-`privateKey` case in the new interface, while we still can.

My reasoning:

- It's very little cost for users to do `{ privateKeys: [privateKey] }` instead of `{ privateKey }`.
  - Naturally, `privateKeys` works perfectly well for both cases, so there's no negative impact except possible confusion.
  - We can remove risk of confusion by specifying "how to pass a single private key" in the README (although I'm not sure it's necessary).
- Including this option adds extra logic:
  - Something has to distinguish this case at run-time (like you've added to `getPrivateKeys`).
  - The type descriptions need to account for the extra possibility, and that gets into risky areas because TS forces you to do either `{ a } | { b }` or `{ a?, b? }`. The latter here is type unsafe; the former adds something else to deal with.

Reply:
I'm not opposed to not supporting `privateKey`. It seems kind of unnecessary but I always hear your voice in the back of my head saying "we should handle/accept as wide a variety of input as possible" haha